Adding accounting monitoring

ivanpavlina opened the pull request:
Working version with updated docs regarding the issue #17 

Brief overview - 
Integration will retrieve and store all DHCP leases and ARP entries, every unique IP address is determined as a new device. After that it builds name for defined IP address, first it tries DHCP lease comment, next is DNS entry for that IP, next is DHCP host-name of the device. If everything fails IP address is set as devices name. Next it retrieves all defined local networks of Mikrotik from DHCP network table which will be used to determine if the traffic is LAN or WAN.
This building and refreshing of devices and names is run every 15 - hardcoded, not sure if this should be exposed as another scan_interval option.

Next, on every update it creates snapshot of accounting data and calculates throughput for given interval (first snapshot is disregarded because that will make spikes in data).

Every sensor (host) has its mac-address and ip-address set as attribute, and its linked by mac-address in integration so for each device there are 2 wan sensors and 2 lan sensors (if the account-local-traffic is enabled in Mikrotik Accounting page).

This is working great from my side, let me know if you want any changes.

ivanpavlina commented:
Yeah, one more thing. There is another entry and error string in translations, so for russian I used google translate :)

tomaae commented on custom_components/mikrotik_router/sensor.py:
```diff
@@ -66,8 +79,49 @@
         ATTR_PATH: "interface",
         ATTR_ATTR: "rx-bits-per-second",
     },
+    "accounting_lan_tx": {
```
Missing ATTR_GROUP for all accounting sensor types.
Not sure where HA would group it in integration page, can you check?

tomaae commented on custom_components/mikrotik_router/sensor.py:
```diff
@@ -276,3 +347,57 @@ def device_info(self):
+                (CONNECTION_NETWORK_MAC, self._data["mac-address"])},
+            "manufacturer": self._ctrl.data["resource"]["platform"],
+            "model": self._ctrl.data["resource"]["board-name"],
+            "name": self._data["name"],
```
This will most likely create group per each accounting name.
Maybe something like "name": self._type[ATTR_GROUP] would be more suitable?
Check how device_info for MikrotikControllerSensor is presented. This could be far more suitable.

tomaae commented on custom_components/mikrotik_router/config_flow.py:
```diff
@@ -131,6 +137,7 @@ def _show_config_form(
                     vol.Optional(CONF_PORT, default=port): int,
                     vol.Optional(CONF_NAME, default=name): str,
                     vol.Optional(CONF_SSL, default=use_ssl): bool,
+                    vol.Optional(CONF_TRACK_ACCOUNTING, default=track_accounting): bool,
```
Adding Enable accounting would be better in integration options like other parameters (in _show_config_form), not in integration setup.

tomaae commented on custom_components/mikrotik_router/sensor.py:
```diff
@@ -276,3 +347,57 @@ def device_info(self):
+    @property
+    def unique_id(self):
+        """Return a unique_id for this entity."""
+        return f"{self._inst.lower()}-{self._sensor.lower()}-{self._data['address'].lower()}"
```
We should use mac-address as unique id, so even when IP changes in case of DHCP, integration will be able to resolve it and pair ip with correct sensor for data history and retain the sensor name.

tomaae commented on custom_components/mikrotik_router/mikrotikapi.py:
```diff
@@ -482,3 +482,106 @@ def get_traffic(self, interfaces) -> Optional(list):
+        return int(round(time() * 1000))
+
+    def is_accounting_enabled(self) -> bool:
+        accounting = self.path("/ip/accounting", return_list=True)
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        # Build hosts from DHCP Server leases and ARP list
+        self.data["accounting"] = parse_api(
+            data=self.data["accounting"],
+            source=self.api.path("/ip/dhcp-server/lease", return_list=True),
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        # Also retrieve all entries in ARP table. If some hosts are missing, build it here
+        arp_hosts = parse_api(
+            data={},
+            source=self.api.path("/ip/arp", return_list=True),
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        #   and then device's host-name. If everything fails use hosts IP address as name
+        dns_data = parse_api(
+            data={},
+            source=self.api.path("/ip/dns/static", return_list=True),
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        # Build list of local networks
+        dhcp_networks = parse_api(
+            data={},
+            source=self.api.path("/ip/dhcp-server/network", return_list=True),
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        if time_diff:
+            accounting_data = parse_api(
+                data={},
+                source=self.api.path("/ip/accounting/snapshot", return_list=True),
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotikapi.py:
```diff
@@ -482,3 +482,106 @@ def get_traffic(self, interfaces) -> Optional(list):
+        return False
+
+    def is_accounting_local_traffic_enabled(self) -> bool:
+        accounting = self.path("/ip/accounting", return_list=True)
```
return_list=True is now default. Since that used the most. only return_list=False have to be specified

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -3,6 +3,7 @@
 import asyncio
 import logging
 from datetime import timedelta
+import ipaddress
```
Do we need full module?

tomaae replied:
only ipaddress.IPv4Network and ipaddress.ip_address are used. 
we choudl use "from ipaddress import IPv4Network, ip_address"

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        for addr in arp_hosts:
+            if addr not in self.data["accounting"]:
+                self.data["accounting"][addr] = {
+                   "address": arp_hosts[addr]['address'],
```
both address and mac-address use arp_hosts[addr]['address']? possible typo

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        self.data["accounting"] = parse_api(
+            data=self.data["accounting"],
+            source=self.api.path("/ip/dhcp-server/lease", return_list=True),
+            key="address",
```
key should be mac-address, to retain history and custom names in HA ++ possibility for automation

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -629,3 +662,180 @@ def get_queue(self):
+        arp_hosts = parse_api(
+            data={},
+            source=self.api.path("/ip/arp", return_list=True),
+            key="address",
```
key should be mac-address, to retain history and custom names in HA ++ possibility for automation

tomaae commented:
Looks really good, I have commented lines where there is possible problem or improvement.
If anything is unclear, just ask away.
Sorry if it looks bit messy, this is first time I used review feature. Ability to comment per line is amazing tho.

tomaae commented:
why remove arp and dns? isnt in necessary to track accounting for hosts with static IP?

ivanpavlina commented:
I will try to reuse existing call to ARP, so no two of the same calls are made in the same update.. Like dhcp-server/lease call

tomaae commented:
ah right. yeah, we should move all api list calls like that to separate get_... function. 
I will separate dhcp from host_tracking too. Same with arp.

having get_* functions to just pull lists from api + processing related just to that list only
and process_* for creating new lists with heavy processing.

it should result in much cleaner and easier to work with code.

ivanpavlina commented:
Okay, great, I am almost finished with this PR, only this left is to remove "Track accounting" option from integration creation. And before metioned ARP call :)

Now I was thinking, maybe it would be best to run "is_accounting_enabled" on startup and every update so entites wouldl be created only if the accounting is enabled, if its not it will skip snapshoting and calculation.
That way the user just enables accounting in Mikrotik and entites are automatically created and there is no need for another configuration by the user.

tomaae commented:
oh yeah, unnecessary api queries and processing should be disabled when the functionality is off in options.
I use it for get_interface_client (option_track_arp). I will need to rename it tho.

you can use something like option_enable_accounting

tomaae commented:
if you want to handle checking if accounting is enabled from router, just be sure to check it from time to time, in case it is enabled or disabled online
also check what happens if you disable accounting when everything is running. just in case it crashes something.

ivanpavlina commented:
Everything is done, if the accounting is never turned on, sensor wont be ever be created. If the accounting is running and it is disabled in Mikrotik all sensors will be defaulted to 0, until accounting is enabled again. I tried all combinations, integration is handling everything pretty well.

Name of entity is always hostname, or DNS static entry name. Or fallback to MAC address if nothing is present.

Regarding the entity_id, it is built by hostname (example sensor.mikrotik_somehostname_wan_tx), not by MAC address, same as DHCP lease device trackers. I think that is okay, it's keeping it inline with current implementation.

I added 'comment' attribute which will be filled with comment from DHCP lease, or if that doesn't exist it will follow the same logic as entity name.

Only need to update the docs a bit and this could be marked as complete, unless you have some ideas?
I just need to update the

tomaae commented:
oh yeah, its perfectly fine to use for entity name. entity unique_id has to use mac address for proper pairing.
For example, I use 'unknown' as last fallback for entity name in tracking right now and mac in unique_id. I have tested that HA will pair it correctly even when I remove every pointer to the device and reconnect it.
unique_id is not visible in HA UI at all anyway, so no need to worry about it not making sense to human eye.

I will review the code with all the changes in a bit.

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+
+        # If some host was added from ARP table build new host-name for it from static DNS entry. Fallback to MAC
+        if host_update_from_arp:
+            dns_data = parse_api(
```
lets change dns_data to self.data["dns_static"] so it can be reused elsewhere

tomaae replied:
and move it out to separate async function get_dns_static

tomaae replied:
available as self.data["dns"]

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+
+        # Build missing hosts from already retrieved ARP list
+        host_update_from_arp = False
+        for entry in self.raw_arp_entries:
```
lets create and use new api_parse in new async function get_arp for arp self.data["arp"] so it can be reused elsewhere.
you can rename current self.data["arp"] I use in get_interface_client to self.data["arp_temp"]. I will redo it later, since its still done in old way and we should have consistency.

tomaae replied:
available as self.data["arp"]
old arp renamed to arp_temp, I will remove it later

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+        )
+
+        # Build list of local DHCP networks
+        dhcp_networks = parse_api(
```
lets create and use new api_parse in new async function get_dhcp_networks for arp self.data["dhcp_networks"] so it can be reused elsewhere.

tomaae replied:
available as self.data["dhcp-network"]

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+                {"name": "address"},
+            ]
+        )
+        self.local_dhcp_networks = [IPv4Network(network) for network in dhcp_networks]
```
it would be probably better to update self.data["dhcp_networks"] within get_dhcp_networks directly instead of adding new variable.
we cant change key, but adding new field should be easy. like:
            ensure_vals=[
                {"name": "IPv4"},
            ]

tomaae replied:
key "IPv4Network" is available as self.data["dhcp-network"]

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+        accounting_enabled, local_traffic_enabled = self.api.is_accounting_and_local_traffic_enabled()
+        traffic_type, traffic_div = self._get_traffic_type_and_div()
+
+        if not accounting_enabled:
```
we could use entity availability here instead using sensor class property.
0 is a value, but if we set entity as unavailable, it will change to gray and wont submit value data to database or automation.
```
    @property
    def available(self) -> bool:
        """Return True if entity is available."""
        return True
```

tomaae replied:
most simple would be probably to store sensor availability in data["accounting"]

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+            return
+
+        # Build missing hosts from already retrieved DHCP Server leases
+        for mac, vals in self.data["dhcp"].items():
```
wont this also add sensors for hosts do not have accounting set up?

tomaae commented on custom_components/mikrotik_router/mikrotik_controller.py:
```diff
@@ -634,3 +665,229 @@ def get_queue(self):
+                        tmp_accounting_values[addr]['lan-tx'] / time_diff * traffic_div, 2)
+                    self.data['accounting'][mac]['lan-rx'] = round(
+                        tmp_accounting_values[addr]['lan-rx'] / time_diff * traffic_div, 2)
+                else:
```
setting them as unavailable here too instead of changing value would be a good idea

tomaae commented:
Just an idea, we could made a single function for resolving both mac address or ip to hostname  which would be used everywhere for simplicity and consistency.
Once we have global get functions for both dns and dhcp, it should be simple.

Priorities could be:
1. DNS static > Name
2. DHCP server leases > Comment
2. DHCP server leases > Active hostname
3. mac address

what do you think?

tomaae commented:
Actually, hold on with changes. I just got some really good idea how to handle hosts.
I will commit them to master during the day. Also I will create data points I mentioned above, so you dont have to.

I will let you know here once its done :)

tomaae commented:
Ok, everything is committed to master. I have added notes to code comments above so you know where to access data.

Also I have created new data point for all hosts, so you no longer need to get them from dhcp and arp separately.
It resolves hostname in same way as you did too.
```
self.data["host"][uid]{ # uid is mac-address
'source': 'arp', # represents data source: dhcp/arp
'address': '10.0.0.1', 
'mac-address': 'XX:XX:XX:XX:XX:XX', 
'interface': 'bridge', 
'hostname': 'somehostname', # resolved hostname, priority: DNS, DHCP comment, DHCP hostname, MAC Address
'last-seen': datetime.datetime(2020, 4, 8, 10, 33, 55, 17106, tzinfo=<UTC>), # time object (util.dt.utcnow)
'available': True # Is accesible right now
}```

ivanpavlina commented:
Awesome work man, this will streamline hosts very much. I'll pull new code and do the changes, I'll let you know

ivanpavlina commented:
> Ok, everything is committed to master. I have added notes to code comments above so you know where to access data.
> 
> Also I have created new data point for all hosts, so you no longer need to get them from dhcp and arp separately.
> It resolves hostname in same way as you did too.
> 
> ```
> self.data["host"][uid]{ # uid is mac-address
> 'source': 'arp', # represents data source: dhcp/arp
> 'address': '10.0.0.1', 
> 'mac-address': 'XX:XX:XX:XX:XX:XX', 
> 'interface': 'bridge', 
> 'hostname': 'somehostname', # resolved hostname, priority: DNS, DHCP comment, DHCP hostname, MAC Address
> 'last-seen': datetime.datetime(2020, 4, 8, 10, 33, 55, 17106, tzinfo=<UTC>), # time object (util.dt.utcnow)
> 'available': True # Is accesible right now
> }```
> ```

Can we build 'hostname' as 'host-name' so it is inline with Mikrotik API?

Also, there was a bug in process_host method, on lines
```
self.data["dhcp"][uid]['interface'] = \
  self.data["dhcp-server"][self.data["dhcp"][uid]['server']]["interface"]

self.data["dhcp"][uid]['available'] = \
  self.api.arp_ping(self.data["dhcp"][uid]['address'], self.data["dhcp"][uid]['interface'])
```
UID wont be present in DHCP dictionary for static ARP entries, I added if statement to check this. See if that is okay. Maybe we could just remove those static ARP entries from hosts dict, atleast in my case there is one so I can broadcast WOL packets to another network - this isn't real host.

As for accounting, thats it, i reused everything that you set up and cleaned up the code from unnecessary variables. I tested it and it works as intended. Check if everything is looking good from your side.

tomaae commented:
`Can we build 'hostname' as 'host-name' so it is inline with Mikrotik API?`
Sure, go ahead and change it. It just needs to be changed everywhere.

`UID wont be present in DHCP dictionary for static ARP entries, I added if statement to check this. See if that is okay. Maybe we could just remove those static ARP entries from hosts dict, atleast in my case there is one so I can broadcast WOL packets to another network - this isn't real host.`
Not sure what you mean.
But that code is different in last commit, maybe you checked old code?
```
self.data["dhcp"][uid]['interface'] = \
  self.data["dhcp-server"][self.data["dhcp"][uid]['server']]["interface"]
```
is called as part of get_dhcp, only for dhcp to fill the interface

```
           if vals["address"] != "unknown" and vals["interface"] != "unknown":
                self.data["host"][uid]["available"] = \
                    self.api.arp_ping(vals["address"], vals["interface"])
```
availability check is called as part of process_host

ivanpavlina commented:
Yep, my bad, i screwed up merge. Sorry, will fix it now.

tomaae commented:
Looks really good :)

ivanpavlina commented:
Damn, i failed with accounting picture, it is saved as .jpg and referenced as .png in Readme. Can you do the quickfix in readme or should I open new PR for this?

tomaae commented:
sure. I fixed both images
